README.md
# 📆 Dateable

[![Pub](https://img.shields.io/pub/v/dateable.svg)](https://pub.dartlang.org/packages/dateable)
<a href="https://www.buymeacoffee.com/jakubkrapiec" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-blue.png" alt="Buy Me A Coffee" height="20"  width="85" style="height: 20px !important;width: 85px !important;"></a>

A Dart package to help you with managing dates easily. Can be used to store, format, convert, construct, parse and serialise dates. Calendar correctness is guaranteed by the usage of `DateTime`'s system under the hood.

## ⚙️ Import

In your `.dart` files:

```dart
import 'package:dateable/dateable.dart';
```

## ⚗️ Usage

### 👷 Constructors

Variety of different constructors allows for great flexibility and interoperability with other types.

```dart
final date = Date(31, 12, 2019);
final date = Date.fromDateTime(DateTime(2019, 12, 31, 19, 1)); // Time of day is truncated
final date = Date.parseIso8601('2019-12-31T18:23:48.956871'); // Time of day is truncated
final date = Date.parse('31122019');
final date = Date.today();
final date = Date.yesterday();
final date = Date.tomorrow();
```

And a handy `DateTime` extension:

```dart
final date = DateTime(2019, 12, 31, 13, 26).toDate(); // Time of day is truncated
```

All of the above result in the same `date` object!

### 📅 Getters

There are 4 getters. Simple and easy.

```dart
final date = Date(11, 3, 2002);
print(date.day); // Prints 11
print(date.month); // Prints 3
print(date.year); // Prints 2002
print(date.weekday) // Prints 1, since it's a Monday
```

### ↔️ Conversion methods

`Date` allows for seamless and easy conversions to most commonly used representations!

```dart
final date = Date(11, 3, 2002);
final dateTime = date.toDateTime(); // Time of day is set to zeros
print(date.toIso8601()); // Prints 2002-03-11T00:00:00.000
print(date.toIso8601(includeTime: false)) // Prints 2002-03-11
print(date.toString()); // Prints 11032002
```

### ✅ Validation

You can validate your dates easily with the `isValid` static method.

Since the default Date constructor doesn't throw on invalid input, and always tries to silently convert to a valid date instead (just like the `DateTime` constructor), you may want to perform the validation explicitly.

```dart
print(Date.isValid(day: 11, month: 3, year: 2002)); // Prints true
print(Date.isValid(day: 29, month: 2, year: 2021)); // Prints false
print(Date.isValid(day: 40, month: 50, year: -99999)); // Prints true
```

### 📊 Comparisions

Comparisions work just like in your well-known `DateTime` objects!

```dart
final earlier = Date(11, 3, 2002);
final later = Date(21, 9, 2004);
print(earlier.isBefore(later)); // True
print(later.isAfter(earlier)); // Also true
```

On top of this, there are also operators `>` (is after) , `<` (is before), `<=`, `>=` and `==`.

Here comes another handy `DateTime` extension:

```dart
DateTime(2002, 3, 11, 14, 56, 28).isTheSameDate(Date(11, 3, 2002));
```

But if all you want is to check if your `Date` is nearby, here you are.

```dart
final date = Date(11, 3, 2002);
date.isToday();
date.isYesterday();
date.isTomorrow();
```

### 📰 Formatting

You can format your `Date`s to `String`s both with top-level constants and with `String` literals:

- `yyyy` - 4 digit year, i.e. 1997
- `yy` - 2 digit year, i.e. 97
- `mm` - 2 digit month, i.e. 03
- `dd` - 2 digit day, i.e. 11

Both of the below options are correct:

```dart
Date(11, 3, 2002).format([dd, '-', mm, '-', yyyy])
```

```dart
Date(11, 3, 2002).format(['dd', '-', 'mm', 'yyyy'])
```

### 🔨 Modifiers

Last but not least, there is a set of useful modifiers. Every `Date` object is immutable by default, so each of them creates a new `Date` object.

```dart
date.addDays(2) == date + 2 // Always true
date.subtractDays(7) == date - 7 // Also always true
```

You can also use the idiomatic copyWith function.

```dart
date.copyWith(day: 21, month: 9);
```

Sorting an `Iterable` of `Date`s chronologically is even easier:

```dart
[Date(21, 9, 2004), Date(24, 12, 2006), Date(11, 3, 2002)].sort((a, b) => a.compareTo(b));
```

Now the list is `[Date(11, 3, 2002), Date(21, 9, 2004), Date(24, 12, 2006)]`.

## 🐛 Contributing / bug reporting

Contributions and bug reports are welcome! Feel free to [open an issue](https://github.com/SugaR256/dateable/issues) or [create a pull request](https://github.com/SugaR256/dateable/pulls).

## 📖 License

This package is distributed under [MIT license](https://github.com/SugaR256/dateable/blob/master/LICENSE).


CHANGELOG.md
## 4.0.1

- Fix too tight dependency constraint.

## 4.0.0

- **BREAKING:** Upgrade minimal Dart version to 2.17.
- Added `weekday` getter.
- Added `includeTime` optional parameter to `toIso8601`.
- Added `isValid` static method.
- Updated dependencies.

## 3.0.0+3

- Improved internal formatting to score full set of pub points.
- Improved internal code clarity.
- Upgraded the internal linting system.
- Get rid of abandoned dependency.

## 3.0.0+2

- Improved performance.
- Added support for package development in Android Studio in .gitignore.
- Introduced stricter internal linting.

## 3.0.0+1

- Improved README formatting.

## 3.0.0

- **BREAKING:** Migrated to NNBD.

## 2.1.1+4

- Upgraded dependencies.

## 2.1.1+3

- Fixed a typo in CHANGELOG.
- Added the Buy Me A Coffee Button to README.
- Updated dependencies.

## 2.1.1+2

- Fixed `meta` dependency version compatibility.

## 2.1.1+1

- Updated `meta` dependency version.

## 2.1.1

- Fixed a typo in README.

## 2.1.0

- Added assertion error messages.
- Fixed typos in documentation.

## 2.0.0

- **BREAKING:** Added `null` check asserts in constructors and methods.
- Added `format` method and top-level constants `dd`, `mm`, `yyyy` and `yy` as its helpers.

## 1.3.0

- Added `today`, `tomorrow` and `yesterday` constructors.

## 1.2.0+1

- Added test coverage badge to README.

## 1.2.0

- Added `isToday` member function.
- Added `isTomorrow` member function.
- Added `isYesterday` member function.

## 1.1.0

- Added `copyWith` member function.
- Added `isTheSameDate` DateTime extension.
- Added `isTheSameDate` member function.

## 1.0.0+4

- Minor correction in CHANGELOG.

## 1.0.0+3

- Minor corrections in README file.

## 1.0.0+2

- Minor corrections in README file.

## 1.0.0+1

- Minor changes to better comply with `pana` requirements.

## 1.0.0

- Initial version.


test/date_test.dart
import 'package:date/date.dart';
import 'package:test/test.dart';

void main() {
  group('Constructor tests.', () {
    test('Date.fromDateTime(DateTime) constructor test.', () {
      expect(
          Date.fromDateTime(DateTime(2002, 3, 11)), equals(Date(11, 3, 2002)));
    });
    test('Date(int, int, int) date validation test.', () {
      expect(Date(50, 40, 2000), equals(Date(20, 5, 2003)));
    });
    test('Date.parseIso8601(String) constructor test.', () {
      expect(Date.parseIso8601('2002-03-11T14:33:21.125'),
          equals(Date(11, 3, 2002)));
    });
    test('Date.parse(String) constructor test.', () {
      expect(Date.parse('11032002'), equals(Date(11, 3, 2002)));
    });
    test('Date.parse(String) constructor exception test.', () {
      var exception;
      try {
        Date.parse('1103207X');
      } catch (e) {
        exception = e;
      }
      expect(exception, isFormatException);
    });
  });
  group('Comparision tests.', () {
    test('compareTo(Date) test: negative.', () {
      expect(Date(11, 3, 2002).compareTo(Date(21, 9, 2004)), equals(-1));
    });
    test('compareTo(Date) test: zero.', () {
      expect(Date(11, 3, 2002).compareTo(Date(11, 3, 2002)), equals(0));
    });
    test('compareTo(Date) test: positive.', () {
      expect(Date(21, 9, 2004).compareTo(Date(11, 3, 2002)), equals(1));
    });
    test('hashCode equality test.', () {
      expect(Date(11, 3, 2002).hashCode, equals(Date(11, 3, 2002).hashCode));
    });
    test('operator== equality test.', () {
      expect(Date(11, 3, 2002) == Date(11, 3, 2002), equals(true));
    });
    test('operator== inequality test.', () {
      expect(Date(11, 3, 2002) == Date(21, 9, 2004), equals(false));
    });
    test('operator> positive test.', () {
      expect(Date(21, 9, 2004) > Date(11, 3, 2002), equals(true));
    });
    test('operator> negative test.', () {
      expect(Date(11, 3, 2002) > Date(21, 9, 2004), equals(false));
    });
    test('operator< positive test.', () {
      expect(Date(11, 3, 2002) < Date(21, 9, 2004), equals(true));
    });
    test('operator< negative test.', () {
      expect(Date(21, 9, 2004) < Date(11, 2, 2002), equals(false));
    });
    test('operator>= positive test.', () {
      expect(Date(11, 3, 2002) >= Date(11, 3, 2002), equals(true));
    });
    test('operator>= negative test.', () {
      expect(Date(11, 3, 2002) >= Date(21, 9, 2004), equals(false));
    });
    test('operator<= positive test.', () {
      expect(Date(11, 3, 2002) <= Date(21, 9, 2004), equals(true));
    });
    test('operator<= negative test.', () {
      expect(Date(21, 9, 2004) <= Date(11, 3, 2002), equals(false));
    });
    test('isBefore(Date) positive test.', () {
      expect(Date(11, 3, 2002).isBefore(Date(21, 9, 2004)), equals(true));
    });
    test('isBefore(Date) negative test,', () {
      expect(Date(11, 3, 2002).isBefore(Date(11, 3, 2002)), equals(false));
    });
    test('isAfter(Date) positive test.', () {
      expect(Date(21, 9, 2004).isAfter(Date(11, 3, 2002)), equals(true));
    });
    test('isAfter(Date) negative test.', () {
      expect(Date(11, 3, 2002).isAfter(Date(11, 3, 2002)), equals(false));
    });
  });
  group('Conversion tests.', () {
    test('toDateTime() test.', () {
      expect(Date(11, 3, 2002).toDateTime(), equals(DateTime(2002, 3, 11)));
    });
    test('toIso8601() test.', () {
      expect(Date(11, 3, 2002).toIso8601(), equals('2002-03-11T00:00:00.000'));
    });
    test('toString() test.', () {
      expect(Date(11, 3, 2002).toString(), equals('11032002'));
    });
    test('toDate() DateTime extension test.', () {
      expect(DateTime(2002, 3, 11).toDate(), equals(Date(11, 3, 2002)));
    });
  });
  group('Addition and subtraction tests.', () {
    test('operator+ test.', () {
      expect(Date(29, 02, 2020) + 1, equals(Date(1, 3, 2020)));
    });
    test('operator- test.', () {
      expect(Date(1, 1, 2020) - 1, equals(Date(31, 12, 2019)));
    });
    test('addDays(int) test.', () {
      expect(Date(29, 02, 2020).addDays(3), equals(Date(3, 3, 2020)));
    });
    test('subtractDays(int) test.', () {
      expect(Date(7, 6, 2020).subtractDays(7), equals(Date(31, 5, 2020)));
    });
  });
}


test/format_test.dart
import 'package:dateable/dateable.dart';
import 'package:test/test.dart';

void main() {
  group('Format tests - predefined constants', () {
    final date = Date(11, 3, 2002);
    test('Format test 1', () {
      expect(date.format([dd, mm, yyyy]), equals('11032002'));
    });
    test('Format test 2', () {
      expect(date.format([dd, '-', mm, '-', yyyy]), equals('11-03-2002'));
    });
    test('Format test 3', () {
      expect(date.format([dd, ' - ', mm, ' - ', yy]), equals('11 - 03 - 02'));
    });
  });
  group('Format tests - by literals', () {
    final date = Date(21, 9, 2004);
    test('Literal test', () {
      expect(date.format(['dd', '/', 'mm', '/', 'yy']), equals('21/09/04'));
    });
    test('Filtering test', () {
      expect(
          date.format(['mmmayyddayyyy, it is ', 'dd', '.', 'mm', '.', 'yyyy']),
          equals('mmmayyddayyyy, it is 21.09.2004'));
    });
  });
}


test/dateable_test.dart
import 'package:dateable/dateable.dart';
import 'package:test/test.dart';

void main() {
  group('Constructor tests', () {
    test('Date.fromDateTime(DateTime) constructor test', () {
      expect(Date.fromDateTime(DateTime(2002, 3, 11)), equals(Date(11, 3, 2002)));
    });
    test('Date(int, int, int) date validation test', () {
      expect(Date(50, 40, 2000), equals(Date(20, 5, 2003)));
    });
    test('Date.parseIso8601(String) constructor test', () {
      expect(Date.parseIso8601('2002-03-11T14:33:21.125'), equals(Date(11, 3, 2002)));
    });
    test('Date.parse(String) constructor test', () {
      expect(Date.parse('11032002'), equals(Date(11, 3, 2002)));
    });
    test('Date.parse(String) constructor exception test', () {
      dynamic exception;
      try {
        Date.parse('1103207X');
      } catch (e) {
        exception = e;
      }
      expect(exception, isFormatException);
    });
    test('today constructor test', () {
      expect(Date.today(), equals(DateTime.now().toDate()));
    });
    test('tomorrow constructor test', () {
      expect(Date.tomorrow(), equals(DateTime.now().add(const Duration(days: 1)).toDate()));
    });
    test('yesterday constructor test', () {
      expect(Date.yesterday(), equals(DateTime.now().subtract(const Duration(days: 1)).toDate()));
    });
  });
  group('Comparision tests', () {
    test('compareTo(Date) test: negative', () {
      expect(Date(11, 3, 2002).compareTo(Date(21, 9, 2004)), equals(-1));
    });
    test('compareTo(Date) test: zero', () {
      expect(Date(11, 3, 2002).compareTo(Date(11, 3, 2002)), equals(0));
    });
    test('compareTo(Date) test: positive', () {
      expect(Date(21, 9, 2004).compareTo(Date(11, 3, 2002)), equals(1));
    });
    test('hashCode equality test', () {
      expect(Date(11, 3, 2002).hashCode, equals(Date(11, 3, 2002).hashCode));
    });
    test('operator== equality test', () {
      expect(Date(11, 3, 2002) == Date(11, 3, 2002), equals(true));
    });
    test('operator== inequality test', () {
      expect(Date(11, 3, 2002) == Date(21, 9, 2004), equals(false));
    });
    test('operator> positive test', () {
      expect(Date(21, 9, 2004) > Date(11, 3, 2002), equals(true));
    });
    test('operator> negative test', () {
      expect(Date(11, 3, 2002) > Date(21, 9, 2004), equals(false));
    });
    test('operator< positive test', () {
      expect(Date(11, 3, 2002) < Date(21, 9, 2004), equals(true));
    });
    test('operator< negative test', () {
      expect(Date(21, 9, 2004) < Date(11, 2, 2002), equals(false));
    });
    test('operator>= positive test', () {
      expect(Date(11, 3, 2002) >= Date(11, 3, 2002), equals(true));
    });
    test('operator>= negative test', () {
      expect(Date(11, 3, 2002) >= Date(21, 9, 2004), equals(false));
    });
    test('operator<= positive test', () {
      expect(Date(11, 3, 2002) <= Date(21, 9, 2004), equals(true));
    });
    test('operator<= negative test', () {
      expect(Date(21, 9, 2004) <= Date(11, 3, 2002), equals(false));
    });
    test('isBefore(Date) positive test', () {
      expect(Date(11, 3, 2002).isBefore(Date(21, 9, 2004)), equals(true));
    });
    test('isBefore(Date) negative test,', () {
      expect(Date(11, 3, 2002).isBefore(Date(11, 3, 2002)), equals(false));
    });
    test('isAfter(Date) positive test', () {
      expect(Date(21, 9, 2004).isAfter(Date(11, 3, 2002)), equals(true));
    });
    test('isAfter(Date) negative test', () {
      expect(Date(11, 3, 2002).isAfter(Date(11, 3, 2002)), equals(false));
    });
    test('copyWith test', () {
      expect(Date(11, 3, 2002).copyWith(day: 21, month: 9), equals(Date(21, 9, 2002)));
    });
    test('isTheSameDate(Date) DateTime extension positive test', () {
      expect(DateTime(2002, 3, 11, 14, 32, 56).isTheSameDate(Date(11, 3, 2002)), equals(true));
    });
    test('isTheSameDate(Date) DateTime extension negative test', () {
      expect(DateTime(2004, 9, 21, 18, 26).isTheSameDate(Date(11, 3, 2002)), equals(false));
    });
    test('isTheSameDate(Date) positive test', () {
      expect(Date(11, 3, 2002).isTheSameDate(Date(11, 3, 2002)), equals(true));
    });
    test('isTheSameDate(Date) negative test', () {
      expect(Date(11, 3, 2002).isTheSameDate(Date(21, 9, 2004)), equals(false));
    });
    test('isToday() positive test', () {
      expect(DateTime.now().toDate().isToday(), equals(true));
    });
    test('isToday() negative test', () {
      expect(DateTime.now().add(const Duration(days: 1)).toDate().isToday(), equals(false));
    });
    test('isTomorrow() positive test', () {
      expect(DateTime.now().add(const Duration(days: 1)).toDate().isTomorrow(), equals(true));
    });
    test('isTomorrow() negative test', () {
      expect(DateTime.now().toDate().isTomorrow(), equals(false));
    });
    test('isYesterday() positive test', () {
      expect(DateTime.now().subtract(const Duration(days: 1)).toDate().isYesterday(), equals(true));
    });
    test('isYesterday() negative test', () {
      expect(DateTime.now().toDate().isYesterday(), equals(false));
    });
  });
  group('Conversion tests', () {
    test('toDateTime() test', () {
      expect(Date(11, 3, 2002).toDateTime(), equals(DateTime(2002, 3, 11)));
    });
    test('toIso8601() with time test', () {
      expect(Date(11, 3, 2002).toIso8601(), equals('2002-03-11T00:00:00.000'));
    });
    test('toIso8601() without time test', () {
      expect(Date(11, 3, 2002).toIso8601(includeTime: false), equals('2002-03-11'));
    });
    test('toString() test', () {
      expect(Date(2, 3, 104).toString(), equals('02030104'));
    });
    test('toDate() DateTime extension test', () {
      expect(DateTime(2002, 3, 11).toDate(), equals(Date(11, 3, 2002)));
    });
  });
  group('Addition and subtraction tests', () {
    test('operator+ test', () {
      expect(Date(29, 02, 2020) + 1, equals(Date(1, 3, 2020)));
    });
    test('operator- test', () {
      expect(Date(1, 1, 2020) - 1, equals(Date(31, 12, 2019)));
    });
    test('addDays(int) test', () {
      expect(Date(29, 02, 2020).addDays(3), equals(Date(3, 3, 2020)));
    });
    test('subtractDays(int) test', () {
      expect(Date(7, 6, 2020).subtractDays(7), equals(Date(31, 5, 2020)));
    });
  });
  group('weekday tests', () {
    test('Weekday getter test 1', () {
      expect(Date(11, 3, 2002).weekday, equals(1));
    });
    test('Weekday getter test 2', () {
      expect(Date(21, 9, 2004).weekday, equals(2));
    });
  });
  group('isValid tests', () {
    test('isValid test 1', () {
      expect(Date.isValid(day: 11, month: 3, year: 2002), equals(true));
    });
    test('isValid test 2', () {
      expect(Date.isValid(day: 29, month: 2, year: 2021), equals(false));
    });
    test('isValid test 3', () {
      expect(Date.isValid(day: 40, month: 50, year: -99999), equals(false));
    });
  });
}


example/date_example.dart
import 'package:date/date.dart';

void main() {
  // The most basic constructor.
  final date = Date(21, 3, 2002);

  // You can also use [fromDateTime] constructor...
  if (date == Date.fromDateTime(DateTime(2002, 3, 21))) {
    print('True!');
  }
  // ...or use an extension...
  if (date == DateTime(2002, 3, 21).toDate()) {
    print('Also true!');
  }
  // ...or parse from [String] ddmmyyyy...
  if (date == Date.parse('21032002')) {
    print('This is awesome.');
  }
  // ...or even parse from an ISO8601 [String]!
  if (date == Date.parseIso8601('2002-03-21T14:35:26.896')) {
    print('Very useful.');
  }

  // Do subtractDays and operator- have the same meaning?
  if (Date(1, 1, 2020) - 2 == Date(1, 1, 2020).subtractDays(2)) {
    print('Yes, they do!');
  }
  // The same with addDays and operator+.
  if (Date(31, 12, 2021) + 18 == Date(1, 1, 2022).addDays(17)) {
    print(':)');
  }
  // You can also use operators >, <, >=, <= and ==...
  if (Date(21, 3, 2002) < Date(21, 9, 2004)) {
    print('Spoiler: it is true.');
  }
  // ...as well as isAfter() and isBefore()!
  if (Date(21, 3, 2002).isBefore(Date(21, 9, 2004))) {
    print('This is also true.');
  }

  // There are also some nice conversions:
  if (Date(21, 3, 2002).toDateTime() == DateTime(2002, 3, 21)) {
    print('Nice.');
  }
  if (Date(21, 3, 2002).toIso8601() == '2002-03-21T00:00:00.000') {
    print('Nice!');
  }
  if (Date(21, 3, 2002).toString() == '21032002') {
    print('<3');
  }

  // Finally, due to implementation of [Comparable] interface,
  // you can sort easily:
  final result = [Date(21, 9, 2004), Date(24, 12, 2006), Date(11, 3, 2002)];
  result.sort((a, b) => a.compareTo(b));
  print(result); // Prints [11032002, 21092004, 24122006]
}


lib/src/dateable_base.dart
import 'package:meta/meta.dart';

/// A class that stores and manages day, month and year.
@immutable
class Date implements Comparable<Date> {
  /// Converts a [DateTime] to a [Date].
  Date.fromDateTime(final DateTime dateTime) : _date = _truncateTimeOfDay(dateTime);

  /// Returns [DateTime] object with the date of [this] and time values at 00:00.
  DateTime toDateTime() => _date;

  /// Returns [0] if [this] is at the same date as [other], [1] if [this] is
  /// after [other] and [-1] if [this] is before [other].
  @override
  int compareTo(final Date other) {
    if (this == other) {
      return 0;
    } else {
      if (_date.isBefore(other.toDateTime())) {
        return -1;
      } else {
        return 1;
      }
    }
  }

  /// Get rid of the given [DateTime]'s time of day data, leaving it all zeroes.
  static DateTime _truncateTimeOfDay(final DateTime dateTime) => DateTime(dateTime.year, dateTime.month, dateTime.day);

  /// Creates a [Date] object from day, month and year values.
  /// If the values are invalid, behaves like [DateTime] constructor, always converting to a valid date.
  /// I.e. [Date(50, 40, 2000)] will be converted to [Date(20, 5, 2003)].
  /// If you want to validate the input, use [Date.isValid] method.
  Date(final int day, final int month, final int year) : _date = DateTime(year, month, day);

  /// Parses given [String] to a new [Date] object. Besides ISO 8601 works with
  /// every format of [String] that [DateTime.parse] would work with.
  /// Performs validation. Throws [FormatException] when the input can't be parsed.
  Date.parseIso8601(final String dateString) : _date = _truncateTimeOfDay(DateTime.parse(dateString));

  /// Returns an ISO8601 [String] representing [this].
  /// If [includeTime] is set to [true], the time will be included, and the format will be "yyyy-mm-ddT00:00:00.000000".
  /// If [includeTime] is set to [false], the time will be excluded, and the format will be "yyyy-mm-dd".
  String toIso8601({final bool includeTime = true}) {
    if (includeTime) {
      return _date.toIso8601String();
    } else {
      return _date.toIso8601String().substring(0, 10);
    }
  }

  /// The [DateTime] object holding the state of [this].
  final DateTime _date;

  /// Parses [String] to [Date] object. [String] must be formatted as ddmmyyyy.
  /// Performs validation. Throws [FormatException] when argument contains non-numbers.
  Date.parse(final String dateString)
      : _date = DateTime(int.parse(dateString.substring(4, 8)), int.parse(dateString.substring(2, 4)),
            int.parse(dateString.substring(0, 2)));

  /// Returns a new [Date] with given amount of days subtracted from [this].
  /// [days] can be negative, in this case addition will happen.
  Date subtractDays(final int days) => _date.subtract(Duration(days: days)).toDate();

  /// Returns a new [Date] with given amount of days added to [this].
  /// [days] can be negative, in this case subtraction will happen.
  Date addDays(final int days) => _date.add(Duration(days: days)).toDate();

  /// Returns a new [Date] with given amount of days added to [this].
  /// [days] can be negative, in this case subtraction will happen.
  Date operator +(final int days) => addDays(days);

  /// Returns a new [Date] with given amount of days subtracted from [this].
  /// [days] can be negative, in this case addition will happen.
  Date operator -(final int days) => subtractDays(days);

  @override
  int get hashCode => toString().hashCode;

  @override
  bool operator ==(final dynamic other) =>
      other is Date && other.day == _date.day && other.month == _date.month && other.year == _date.year;

  /// Checks if [this] is after [other].
  bool operator >(final Date other) => _date.isAfter(other.toDateTime());

  /// Checks if [this] is before [other].
  bool operator <(final Date other) => _date.isBefore(other.toDateTime());

  /// Checks if [this] is after or at the same day as [other].
  bool operator >=(final Date other) => this == other || this > other;

  /// Checks if [this] is before or at the same day as [other].
  bool operator <=(final Date other) => this == other || this < other;

  /// Checks if [this] is before [other].
  bool isBefore(final Date other) => this < other;

  /// Checks if [this] is after [other].
  bool isAfter(final Date other) => this > other;

  /// Checks if [this] is the same date as [other].
  bool isTheSameDate(final Date other) => _date.isTheSameDate(other);

  /// Returns today's date.
  Date.today() : _date = _truncateTimeOfDay(DateTime.now());

  /// Returns tomorrow's date.
  Date.tomorrow() : _date = _truncateTimeOfDay(DateTime.now().add(const Duration(days: 1)));

  /// Returns yesterday's date.
  Date.yesterday() : _date = _truncateTimeOfDay(DateTime.now().subtract(const Duration(days: 1)));

  /// Returns day of the date represented by this object. Always in range [1; 31].
  int get day => _date.day;

  /// Returns month of the date represented by this object. Always in range [1; 12].
  int get month => _date.month;

  /// Returns year of the date represented by this object.
  int get year => _date.year;

  /// Returns weekday of the date represented by this object. Always in range [1; 7], where 1 is Monday and 7 is Sunday.
  int get weekday => _date.weekday;

  /// Returns this [Date]'s object representations as [String], with
  /// format ddmmyyyy. For other formatting options, use [format] method.
  @override
  String toString() {
    final day = _date.day.toString();
    final month = _date.month.toString();
    final year = _date.year.toString();
    return '${_modifyLength(day, 2)}${_modifyLength(month, 2)}${_modifyLength(year, 4)}';
  }

  /// Modified the length of the [digits]. If original length is shorter than [designatedLength],
  /// appropriate amount of 0s are added at the beginning. If original length is shorter than
  /// [designatedLength], front digits of [digits] are truncated.
  String _modifyLength(final String digits, final int designatedLength) {
    final difference = designatedLength - digits.length;
    if (difference < 0) {
      return digits.substring(difference.abs());
    } else {
      return digits.padLeft(designatedLength, '0');
    }
  }

  /// Returns new [Date] modified with given input.
  Date copyWith({final int? day, final int? month, final int? year}) =>
      Date(day ?? _date.day, month ?? _date.month, year ?? _date.year);

  /// Checks if [this] is today.
  bool isToday() => _date.isTheSameDate(Date.today());

  /// Checks if [this] is yesterday.
  bool isYesterday() => _date.isTheSameDate(Date.today() - 1);

  /// Checks if [this] is tomorrow.
  bool isTomorrow() => _date.isTheSameDate(Date.today() + 1);

  /// Formats [this] with a given input.
  /// I.e. to get the [String] `"11-03-2002"`, you want to call `Date(11, 3, 2002).format([dd, '-', mm, '-', yyyy])`.
  /// Available date modifiers are:
  /// * `yyyy` - 4 digit year, i.e. 2002.
  /// * `yy` - 2 digit year, i.e. 02.
  /// * `mm` - 2 digit month, i.e. 03.
  /// * `dd` - 2 digit day, i.e. 11.
  ///
  /// Everything else will be left without change.
  /// You can use all of them both as "dd" and dd.
  String format(final List<String> formats) {
    final result = StringBuffer();
    for (final format in formats) {
      switch (format) {
        case yyyy:
          result.write(_modifyLength(_date.year.toString(), 4));
          break;
        case yy:
          result.write(_modifyLength(_date.year.toString(), 2));
          break;
        case dd:
          result.write(_modifyLength(_date.day.toString(), 2));
          break;
        case mm:
          result.write(_modifyLength(_date.month.toString(), 2));
          break;
        default:
          result.write(format);
          break;
      }
    }
    return result.toString();
  }

  static bool isValid({required final int day, required final int month, required final int year}) {
    final date = DateTime(year, month, day);
    return date.day == day && date.month == month && date.year == year;
  }
}

/// 2 digit month representation, i.e. 03.
const String mm = 'mm';

/// 2 digit day representation, i.e. 11.
const String dd = 'dd';

///2 digit year representation, i.e. 02.
const String yy = 'yy';

/// 4 digit year representation, i.e. 2002.
const String yyyy = 'yyyy';

/// A set of extensions to work with [DateTime]s more seamlessly.
extension DateExtensions on DateTime {
  /// Converts [this] to a corresponding [Date] object.
  Date toDate() => Date.fromDateTime(this);

  /// Checks is [this] is on the same date as [other].
  bool isTheSameDate(final Date other) => toDate() == other;
}


example/dateable_example.dart
import 'package:dateable/dateable.dart';

void main() {
  // The most basic constructor.
  final date = Date(21, 3, 2002);

  // You can also use [fromDateTime] constructor...
  if (date == Date.fromDateTime(DateTime(2002, 3, 21))) {
    print('True!');
  }

  // ...or use an extension...
  if (date == DateTime(2002, 3, 21).toDate()) {
    print('Also true!');
  }

  // ...or parse from [String] ddmmyyyy...
  if (date == Date.parse('21032002')) {
    print('This is awesome.');
  }

  // ...or even parse from an ISO8601 [String]!
  if (date == Date.parseIso8601('2002-03-21T14:35:26.896')) {
    print('Very useful.');
  }

  // Do subtractDays and operator- have the same meaning?
  if (Date(1, 1, 2020) - 2 == Date(1, 1, 2020).subtractDays(2)) {
    print('Yes, they do!');
  }

  // The same with addDays and operator+.
  if (Date(31, 12, 2021) + 18 == Date(1, 1, 2022).addDays(17)) {
    print(':)');
  }

  // You can also use operators >, <, >=, <= and ==...
  if (Date(21, 3, 2002) < Date(21, 9, 2004)) {
    print('Spoiler: it is true.');
  }

  // ...as well as isAfter() and isBefore()!
  if (Date(21, 3, 2002).isBefore(Date(21, 9, 2004))) {
    print('This is also true.');
  }

  // Checking if [DateTime] is on the same day as your [Date]
  // is incredibly easy.
  if (DateTime(2002, 3, 11, 14, 6).isTheSameDate(Date(11, 3, 2002))) {
    print('A really nice extension.');
  }

  // Checking if the [Date] is today, tomorrow or yesterday is even easier.
  if (date.isToday() || date.isTomorrow() || date.isYesterday()) {
    print('The star is born.');
  }

  // You also gain access to [today], [tomorrow] and [yesterday] constructors.
  print(Date.yesterday().toString() +
      Date.today().toString() +
      Date.tomorrow().toString());

  // There are also some nice conversions:
  if (Date(21, 3, 2002).toDateTime() == DateTime(2002, 3, 21)) {
    print('Nice.');
  }
  if (Date(21, 3, 2002).toIso8601() == '2002-03-21T00:00:00.000') {
    print('Nice!');
  }
  if (Date(21, 3, 2002).toString() == '21032002') {
    print('<3');
  }

  // You can enjoy the copyWith function!
  if (Date(11, 3, 2002).copyWith(day: 21, month: 9) == Date(21, 9, 2002)) {
    print('Useful for short, idiomatic mutations.');
  }

  // Formatting is as easy as it gets!
  print(date.format([dd, '-', mm, '-', yyyy]));

  // Finally, due to the implementation of [Comparable] interface,
  // you can sort easily:
  final result = [Date(21, 9, 2004), Date(24, 12, 2006), Date(11, 3, 2002)];
  result.sort((a, b) => a.compareTo(b));
  print(result); // Prints [11032002, 21092004, 24122006]
}
